bundles/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/internal/HarmonyHubBinding.java

```diff
@@ -0,0 +1,243 @@
+						}
+					};
+					harmonyHubGateway.addListener(config.getQualifier(), listener);
+					harmonyListeners.put(config.getQualifier(), listener);
+				}
+			}
+		}
+	}
+
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void allBindingsChanged(BindingProvider provider) {
+		for(String itemName : provider.getItemNames()) {
+			bindingChanged(provider,itemName);
+		}
+	}
+
+	/**
+	 * @{inheritDoc}
+	 */
+	@Override
+	protected void internalReceiveCommand(String itemName, Command command) {
+		logger.debug("internalReceiveCommand({},{}) is called!", itemName, command);
+
+		if(harmonyHubGateway == null){
+			logger.warn("A command was receieved, but could not be executed as no Harmony Hub has been configured");
+			return;
+		}
+
+		for (HarmonyHubBindingProvider provider : providers) {
+			HarmonyHubBindingConfig config = provider.getHarmonyHubBindingConfig(itemName);
+			
+			if(config.getBindingType().getDirection() == HarmonyHubBindingDirection.IN){
+				logger.warn("item {} is not configured to send outbound commands! Please change to > or * in the binding");
+				continue;
+			}
+			
+			switch(config.getBindingType()) {
+			case PressButton:
+				logger.debug("PressButton command:{} q:{} p1:{} p2:{}", command.toString(), config.getQualifier(),
+						config.getParam1(), config.getParam2());
+				String cmd = config.getParam2();
+				if(cmd == null){
+					cmd = command.toString();
+				}
+				try {
+					harmonyHubGateway.pressButton(config.getQualifier(), Integer.parseInt(config.getParam1()), cmd);
+				} catch (NumberFormatException ignored) {
+					harmonyHubGateway.pressButton(config.getQualifier(), config.getParam1(), cmd);
+				}
+				break;
+			case StartActivity:
+				try {
+					harmonyHubGateway.startActivity(config.getQualifier(), Integer.parseInt(config.getParam1()));
+				} catch (NumberFormatException ignored) {
+					harmonyHubGateway.startActivity(config.getQualifier(), config.getParam1());
+				}
+				break;
+			case CurrentActivity:
+				harmonyHubGateway.startActivity(config.getQualifier(),command.toString());
+				break;
+			default :
+				break;
+			}
+		}
+	}
+
+	/**
+	 * For a given {@link Activity}, update all items who need to know about it
+	 * @param activity
+	 */
+	private void updateActivity(Activity activity, String qualifier) {
+		logger.debug("updateActivity {}" + activity.getLabel());
+		for (HarmonyHubBindingProvider provider : providers) {
+			for(String itemName : provider.getItemNames()) {
+				Item item = provider.getItem(itemName);
+				HarmonyHubBindingConfig config = provider.getHarmonyHubBindingConfig(itemName);
+				if(config.matchesQualifier(qualifier) && 
+						config.getBindingType() == HarmonyHubBindingType.CurrentActivity) {
+					updateActivityForItem(item, activity);
+				}
+			}
+		}
+	}
+	
+	/**
+	 * For a given {@link Activity}, update an {@link Item} state
+	 * @param item
+	 * @param activity
+	 */
+	private void updateActivityForItem(Item item , Activity activity) {
```

Comment:
Dan, you might be able to eliminate it by stashing the Item's _name_ (String) and any _type_ information into your `HarmonyHubBindingConfig` (and keying that in your _items_ `Set` by Item's _name_) since you build those up front, at a point where OH has pushed you the `Item` as a parameter, it's possible for your Binding Config class to capture _just_ those attributes of the `Item`, and not maintain a handle to to the Item (it's, state, etc, which I think is what Thomas is getting at)

When you need the Item's typing information, you can get it back from the `HarmonyHubBindingConfig` object (which you can get readily from the caller), and extract the Item typing information to make the branch-test logic.

Of course, I'm simplifying things, as I haven't looked at all the code referencing the _items_ Set...